contracts/ModifiedTokenDistro.sol

```diff
@@ -0,0 +1,339 @@
+     *
+     */
+    function assign(address distributor, uint256 amount) external override {
+        require(hasRole(DEFAULT_ADMIN_ROLE, msg.sender), 'TokenDistro::assign: ONLY_ADMIN_ROLE');
+        require(hasRole(DISTRIBUTOR_ROLE, distributor), 'TokenDistro::assign: ONLY_TO_DISTRIBUTOR_ROLE');
+
+        balances[address(this)].allocatedTokens = balances[address(this)].allocatedTokens - amount;
+        balances[distributor].allocatedTokens = balances[distributor].allocatedTokens + amount;
+
+        emit Assign(msg.sender, distributor, amount);
+    }
+
+    /**
+     * Function to claim tokens for a specific address. It uses the current timestamp
+     *
+     * Emits a {claim} event.
+     *
+     */
+    function claimTo(address account) external {
+        // This check is not necessary as it does not break anything, just changes the claimed value
+        // for this contract
+        //require(address(this) != account, "TokenDistro::claimTo: CANNOT_CLAIM_FOR_CONTRACT_ITSELF");
+        _claim(account);
+    }
+
+    /**
+     * Function to claim tokens for a specific address. It uses the current timestamp
+     *
+     * Emits a {claim} event.
+     *
+     */
+    function claim() external override {
+        _claim(msg.sender);
+    }
+
+    /**
+     * Function that allows to the distributor address to allocate some amount of tokens to a specific recipient
+     * @param recipient of token allocation
+     * @param amount allocated amount
+     * @param claim whether claim after allocate
+     *
+     * Emits a {Allocate} event.
+     *
+     */
+    function _allocate(address recipient, uint256 amount, bool claim) internal {
+        require(!hasRole(DISTRIBUTOR_ROLE, recipient), 'TokenDistro::allocate: DISTRIBUTOR_NOT_VALID_RECIPIENT');
+
+        balances[msg.sender].allocatedTokens = balances[msg.sender].allocatedTokens - amount;
+
+        balances[recipient].allocatedTokens = balances[recipient].allocatedTokens + amount;
+
+        if (claim && claimableNow(recipient) > 0) {
+            _claim(recipient);
+        }
+
+        emit Allocate(msg.sender, recipient, amount);
+    }
+
+    function allocate(address recipient, uint256 amount, bool claim) external override onlyDistributor {
+        _allocate(recipient, amount, claim);
+    }
+
+    /**
+     * Function that allows to the distributor address to allocate some amounts of tokens to specific recipients
+     * @dev Needs to be initialized: Nobody has the DEFAULT_ADMIN_ROLE and all available tokens have been assigned
+     * @param recipients of token allocation
+     * @param amounts allocated amount
+     *
+     * Unlike allocate method it doesn't claim recipients available balance
+     */
+    function _allocateMany(address[] memory recipients, uint256[] memory amounts) internal onlyDistributor {
+        require(recipients.length == amounts.length, 'TokenDistro::allocateMany: INPUT_LENGTH_NOT_MATCH');
+
+        for (uint256 i = 0; i < recipients.length; i++) {
```

Comment:
lets cache array length before for loop and then use it. will save gas.


uint256 length = recipients.length;
for (uint256 i = 0; i < length; i++) {